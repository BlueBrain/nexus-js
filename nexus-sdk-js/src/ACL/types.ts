export type Context = string | string[];

export type IdentityType = 'User' | 'Group' | 'Authenticated' | 'Anonymous';

export interface Identity {
  '@id': string;
  '@type': IdentityType;
  subject?: string;
  realm?: string;
  group?: string;
}

export interface ACLResponseCommon {
  '@id': string;
  '@type': IdentityType;
  _path: string;
  acl: {
    permissions: string[];
    identity: Identity;
  }[];
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
  _rev: number;
}

export interface ACLResponse extends ACLResponseCommon {
  '@context': Context;
}

export interface ListACLResponse {
  '@context': Context;
  _total: number;
  _results: ACLResponseCommon[];
}

export interface ACLOperationSuccess {
  '@context': Context;
  '@id': string;
  '@type': 'AccessControlList';
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
  _rev: number;
}

export interface ListRealmOption {
  from?: number;
  size?: number;
  rev?: number;
  ancestors?: true;
  self?: boolean;
  [key: string]: any;
}

export interface ACLPayload {
  permissions: string[];
  identity: {
    subject?: string;
    realm?: string;
    group?: string;
  };
}
