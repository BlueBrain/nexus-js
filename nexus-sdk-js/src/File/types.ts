import { Context } from '../Resource/types';

export interface FileResponse {
  '@context': Context;
  '@id': string;
  '@type': 'File';
  _bytes: number;
  _digest: {
    _algorithm: string;
    _value: string;
  };
  _filename: string;
  _mediaType: string;
  _self: string;
  _constrainedBy: string;
  _project: string;
  _rev: 1;
  _deprecated: false;
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
}
