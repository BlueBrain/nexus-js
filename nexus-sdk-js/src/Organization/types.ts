export interface ListOrgOptions {
  full_text_search?: string;
  from?: number;
  size?: number;
  deprecated?: boolean;
  [key: string]: any;
}

export interface CreatOrgPayload {
  description?: string;
}
export type Context = string | string[];
