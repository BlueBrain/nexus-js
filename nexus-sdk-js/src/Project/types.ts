export interface ProjectResponseCommon {
  '@id': string;
  '@type': string;
  base: string;
  vocab: string;
  apiMappings: ApiMapping[];
  _label: string;
  _organizationLabel: string;
  _organizationUuid: string;
  _uuid: string;
  _rev: number;
  _deprecated: boolean;
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
  description?: string;
}

export interface ListProjectsResponse {
  _total: number;
  _links?: any;
  _results?: ProjectResponse[];
  '@context'?: Context;
  code?: string;
  message?: string;
}

export interface ProjectResponse extends ProjectResponseCommon {
  '@context'?: Context;
}

export interface ApiMapping {
  prefix: string;
  namespace: string;
}

export interface CreateProjectPayload {
  description?: string;
  base?: string;
  vocab?: string;
  apiMappings?: ApiMapping[];
}

export interface ListProjectOptions {
  q?: string;
  from?: number;
  size?: number;
  deprecated?: boolean;
  [option: string]: any;
}

export type Context = string | string[];

export type ProjectEventType =
  | 'ProjectCreated'
  | 'ProjectUpdated'
  | 'ProjectDeprecated';

export interface ProjectEvent {
  '@context': Context;
  '@type': ProjectEventType;
  apiMappings: ApiMapping[];
  base: string;
  description: string;
  vocab: string;
  _instant: string;
  _label: string;
  _rev: number;
  _subject: string;
  _uuid: string;
}

export interface ProjectCreatedEvent extends ProjectEvent {
  '@type': 'ProjectCreated';
}
export interface ProjectUpdatedEvent extends ProjectEvent {
  '@type': 'ProjectCreated';
}
export interface ProjectDeprecatedEvent extends ProjectEvent {
  '@type': 'ProjectCreated';
}

export interface ProjectEventListeners {
  onOpen?(): void;
  onError?(): void;
  onProjectCreated?(event: ProjectCreatedEvent): void;
  onProjectUpdated?(event: ProjectUpdatedEvent): void;
  onProjectDeprecated?(event: ProjectDeprecatedEvent): void;
}
