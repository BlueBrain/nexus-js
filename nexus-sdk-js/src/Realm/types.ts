import { PaginationSettings } from '../utils/types';

export type Context = string | string[];

export interface RealmResponseCommon {
  '@id': string;
  '@type': string;
  name: string;
  openidConfig: string;
  _label: string;
  _issuer: string;
  _authorizationEndpoint: string;
  _tokenEndpoint: string;
  _userInfoEndpoint: string;
  _endSessionEndpoint: string;
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
  _rev: number;
  _deprecated: boolean;
}
export interface RealmResponse extends RealmResponseCommon {
  '@context': Context;
}
export interface ListRealmResponse {
  '@context': Context;
  _total: number;
  _results: RealmResponseCommon[];
}

export interface ListRealmOptions extends PaginationSettings {
  deprecated?: boolean;
  [key: string]: any;
}
