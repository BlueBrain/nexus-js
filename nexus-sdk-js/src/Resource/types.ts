export type Context = string | object;

export interface ListResourceResponseError {
  '@context': Context;
  code: string;
  message: string;
}

export interface ResourceResponseError {
  '@context': Context;
  code: string;
  message: string;
  ref: string;
}

export interface ResourceListTagResponse {
  '@context': Context;
  tags: string[];
}

export interface CreateResourcePayload {
  resourceId?: string;
  type?: string[];
  context: { [field: string]: string };
  [field: string]: any;
}
