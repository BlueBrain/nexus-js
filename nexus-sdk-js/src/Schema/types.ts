export type Context = string | object;

export type Shape = {
  '@id': string;
  '@type': string;
  nodeKing: string;
  property: {
    datatype: string;
    minCount: number;
    path: string;
  };
  targetClass: string;
};

export interface SchemaResponseCommon {
  '@id': string;
  '@type': 'Schema';
  shapes: Shape[];
  _self: string;
  _constrainedBy: string;
  _project: string;
  _createdAt: string;
  _createdBy: string;
  _updatedAt: string;
  _updatedBy: string;
  _rev: number;
  _deprecated: boolean;
}

export interface ListSchemaResponse {
  '@context': Context;
  _total: number;
  _results: SchemaResponseCommon[];
}

export interface SchemaResponse extends SchemaResponseCommon {
  '@context': Context;
}

export interface ListSchemaResponseError {
  '@context': Context;
  code: string;
  message: string;
}

export interface SchemaResponseError {
  '@context': Context;
  code: string;
  message: string;
  ref: string;
}

export interface SchemaListTagResponse {
  '@context': Context;
  tags: string[];
}

export interface CreateSchemaPayload {
  schemaId?: string;
  context: { [field: string]: string };
  shapes: Shape[];
  [field: string]: any;
}

export interface ListSchemaOptions {
  full_text_search?: string;
  from?: number;
  size?: number;
  deprecated?: boolean;
  rev?: number;
  type?: string;
  createdBy?: string;
  updatedBy?: string;
  [key: string]: any;
}
