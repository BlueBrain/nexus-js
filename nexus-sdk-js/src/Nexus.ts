import Store from './utils/Store';
import Organization, { OrganizationResponse } from './Organization';
import { httpGet, httpPost } from './utils/http';

type NexusConfig = {
  environment: string;
  token?: string;
};

export const store: Store = new Store({
  auth: {
    accessToken: null,
  },
  api: {
    baseUrl: null,
  },
});

export default class Nexus {
  constructor(config: NexusConfig) {
    if (!config.environment) {
      throw new Error(
        'No environment provided. Please specify your Nexus instance endpoint.'
      );
    }
    store.update('api', state => ({
      ...state,
      baseUrl: config.environment,
    }));
    if (config.token) {
      store.update('auth', state => ({
        ...state,
        accessToken: config.token,
      }));
    }
  }

  setToken(token: string): void {
    store.update('auth', state => ({
      ...state,
      accessToken: token,
    }));
  }

  removeToken(): void {
    store.update('auth', state => ({
      ...state,
      accessToken: undefined,
    }));
  }

  async listOrganizations(): Promise<Organization[]> {
    try {
      const orgs: OrganizationResponse[] = await httpGet('/orgs');
      return orgs.map((org: OrganizationResponse) => new Organization(org));
    } catch (e) {
      console.log(e);
      return e;
    }
  }

  async getOrganization(id: string): Promise<Organization> {
    try {
      const org: OrganizationResponse = await httpGet(`/orgs/${id}`);
      return new Organization(org);
    } catch (e) {
      return e;
    }
  }

  async createOrganization(): Promise<Organization> {
    try {
      const org: OrganizationResponse = await httpPost('/orgs', {});
      return new Organization(org);
    } catch (e) {
      return e;
    }
  }
}
