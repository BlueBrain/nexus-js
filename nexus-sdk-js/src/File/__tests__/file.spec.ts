import { resetMocks, mock, mockResponse } from 'jest-fetch-mock';
import Nexus, { File as NexusFile } from '../..';
import { FileResponse } from '../types';
import { Readable } from 'stream';
import fs from 'fs';

const baseUrl = 'http://api.url';
Nexus.setEnvironment(baseUrl);

const mockPostFileResponse: FileResponse = {
  '@context': 'https://bluebrain.github.io/nexus/contexts/resource.json',
  '@id': 'base:d8848d4c-68f7-4ffd-952f-63a8cbcb86a9',
  '@type': 'File',
  _bytes: 670,
  _digest: {
    _algorithm: 'SHA-256',
    _value: '25fc54fba0beec17a598b5a68420ded1595b2f76f0a0b7c6077792ece828bc2e',
  },
  _filename: 'myfile.jpg',
  _mediaType: 'image/jpeg',
  _self:
    'https://nexus.example.com/v1/files/myorg/myproj/base:d8848d4c-68f7-4ffd-952f-63a8cbcb86a9',
  _constrainedBy: 'https://bluebrain.github.io/nexus/schemas/file.json',
  _project: 'https://nexus.example.com/v1/projects/myorg/myproj',
  _rev: 1,
  _deprecated: false,
  _createdAt: '2019-01-28T12:15:33.238Z',
  _createdBy: 'https://nexus.example.com/v1/anonymous',
  _updatedAt: '2019-01-28T12:15:33.238Z',
  _updatedBy: 'https://nexus.example.com/v1/anonymous',
};

describe('File class', () => {
  afterEach(() => {
    resetMocks();
  });
  describe('It should create a File instance', () => {
    it('from a getByID example response', () => {
      const resource = new NexusFile(
        'testOrg',
        'testProject',
        mockPostFileResponse,
      );
      expect(resource).toMatchSnapshot();
    });
  });

  describe('File.create()', () => {
    it('should POST the new file with the expected payload', async () => {
      mockResponse(JSON.stringify(mockPostFileResponse), { status: 200 });
      const buffer = new Buffer('abc');
      const stream = new Readable();
      stream.push(buffer);
      stream.push(null);

      await NexusFile.create('myOrg', 'myProject', stream);
      const body = mock.calls[0][1].body;
      expect(body._overheadLength).toBe(143);
    });
  });
});
