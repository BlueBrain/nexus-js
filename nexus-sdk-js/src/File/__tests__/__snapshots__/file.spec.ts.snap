// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File class It should create a File instance from a getByID example response 1`] = `
NexusFile {
  "bytes": 670,
  "constrainedBy": "https://bluebrain.github.io/nexus/schemas/file.json",
  "context": "https://bluebrain.github.io/nexus/contexts/resource.json",
  "createdAt": "2019-01-28T12:15:33.238Z",
  "createdBy": "https://nexus.example.com/v1/anonymous",
  "deprecated": false,
  "digest": Object {
    "algorithm": "SHA-256",
    "value": "25fc54fba0beec17a598b5a68420ded1595b2f76f0a0b7c6077792ece828bc2e",
  },
  "filename": "myfile.jpg",
  "id": "base:d8848d4c-68f7-4ffd-952f-63a8cbcb86a9",
  "mediaType": "image/jpeg",
  "orgLabel": "testOrg",
  "project": "https://nexus.example.com/v1/projects/myorg/myproj",
  "projectLabel": "testProject",
  "raw": Object {
    "@context": "https://bluebrain.github.io/nexus/contexts/resource.json",
    "@id": "base:d8848d4c-68f7-4ffd-952f-63a8cbcb86a9",
    "@type": "File",
    "_bytes": 670,
    "_constrainedBy": "https://bluebrain.github.io/nexus/schemas/file.json",
    "_createdAt": "2019-01-28T12:15:33.238Z",
    "_createdBy": "https://nexus.example.com/v1/anonymous",
    "_deprecated": false,
    "_digest": Object {
      "_algorithm": "SHA-256",
      "_value": "25fc54fba0beec17a598b5a68420ded1595b2f76f0a0b7c6077792ece828bc2e",
    },
    "_filename": "myfile.jpg",
    "_mediaType": "image/jpeg",
    "_project": "https://nexus.example.com/v1/projects/myorg/myproj",
    "_rev": 1,
    "_self": "https://nexus.example.com/v1/files/myorg/myproj/base:d8848d4c-68f7-4ffd-952f-63a8cbcb86a9",
    "_updatedAt": "2019-01-28T12:15:33.238Z",
    "_updatedBy": "https://nexus.example.com/v1/anonymous",
  },
  "rev": 1,
  "self": "https://nexus.example.com/v1/files/myorg/myproj/base:d8848d4c-68f7-4ffd-952f-63a8cbcb86a9",
  "type": "File",
  "updatedAt": "2019-01-28T12:15:33.238Z",
  "updatedBy": "https://nexus.example.com/v1/anonymous",
}
`;
